tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -65,6 +77,86 @@ struct CapabilityDef : public RefObject
             return nullptr;
         return expr.conjunctions[0].atoms[0];
     }
+
+    const HashSet<CapabilityDef*>& getAbstractAtomsPresent() const
+    {
+        return abstractAtomsPresent;
+    }
+    void fillAbstractAtomsPresentFromCannonicalRepresentation()
+    {
+        bool alreadySetTarget = false;
+        HashSet<CapabilityDef*> firstTargetSet;
+
+        bool alreadySetStage = false;
+        HashSet<CapabilityDef*> firstStageSet;
+
+        HashSet<CapabilityDef*> abstractFound;
+
+        for (auto& c : canonicalRepresentation)
+        {
+
+            alreadySetTarget = false;
+            firstTargetSet.clear();
+            alreadySetStage = false;
+            firstStageSet.clear();
+
+            abstractFound.clear();
+
+            for (auto& atom : c)
+            {
+                for (auto& otherAbstractAtomsPresent : atom->abstractAtomsPresent)
+                {
+                    auto base = otherAbstractAtomsPresent->getAbstractBase();
+                    if (CapabilityDef::ptrOfTarget && base == CapabilityDef::ptrOfTarget)
+                    {
+                        if (!alreadySetTarget)
+                            firstTargetSet.add(otherAbstractAtomsPresent);
+                    }
+                    else if (CapabilityDef::ptrOfStage && base == CapabilityDef::ptrOfStage)
+                    {
+                        if (!alreadySetStage)
+                            firstStageSet.add(otherAbstractAtomsPresent);
+                    }
+
+                    abstractFound.add(otherAbstractAtomsPresent);
+                }
```

Comment:
I am not sure about the performance but the logic will be much simpler if you break this loop into two loops.
```
for (auto otherAbstractAtomsPresent : atom->abstractAtomsPresent)
    abstractFound.add(otherAbstractAtomsPresent);

for (auto& otherAbstractAtomsPresent : atom->abstractAtomsPresent)
{
    auto base = otherAbstractAtomsPresent->getAbstractBase();

    if (!alreadySetTarget)
    {
        if (CapabilityDef::ptrOfTarget && base == CapabilityDef::ptrOfTarget)
        {
            for (auto item : otherAbstractAtomsPresent)
                if (abstractFound.contains(item))
                {
                    firstTargetSet.add(item);
                    alreadySetTarget = true;
                }
        }
    }

    if (!alreadySetStage)
    {
        if (CapabilityDef::ptrOfStage && base == CapabilityDef::ptrOfStage)
        {
            for (auto item : otherAbstractAtomsPresent)
                if (!abstractFound.contains(item))
                {
                    firstStageSet.add(item);
                    alreadySetTarget = true;
                }
        }
    }
````

Reply:
I will change the function to be more neat.